syntax = "proto3";

package coolbox.toolbus;

option java_package = "com.coolbox.toolbus";

message Header {
  string request_id = 1;
  string tool = 2;
  map<string, string> metadata = 3;
}

enum StatusCode {
  STATUS_OK = 0;
  STATUS_ERROR = 1;
  STATUS_NOT_FOUND = 2;
  STATUS_UNAVAILABLE = 3;
  STATUS_GUARD_REJECTED = 4;
}

message InvokeRequest {
  Header header = 1;
  bytes payload = 2;
}

message InvokeResponse {
  string request_id = 1;
  StatusCode status = 2;
  bytes payload = 3;
  string error = 4;
}

message StreamRequest {
  Header header = 1;
  bytes payload = 2;
}

message StreamChunk {
  string request_id = 1;
  bytes payload = 2;
  bool end_of_stream = 3;
  StatusCode status = 4;
  string error = 5;
}

message SubscribeRequest {
  Header header = 1;
  repeated string topics = 2;
}

message Event {
  string topic = 1;
  bytes payload = 2;
  map<string, string> metadata = 3;
  string request_id = 4;
}

message Ack {
  string request_id = 1;
  StatusCode status = 2;
  string error = 3;
}
